### PR TITLE
MWPW-159257 refactor mas enablement

### PR DESCRIPTION
**Pull request by npeltier**
- move of wcms-commerce -> mas-commerce-service
- move of some public functions to mas-commerce-service,
- remove mas.js code for initialization, everything needs to go through the component,
- split of settings code with locale settings, make milo locale settings an explicit function,
- unit tests,
- updates of docs, especially  mas.js.html with that component

Resolves: [MWPW-159257](https://jira.corp.adobe.com/browse/MWPW-159257)

**Test URLs:**
- Before: https://main--milo--adobecom.hlx.page/libs/features/mas/docs/mas.js.html
- After: https://MWPW-159257--milo--npeltier.hlx.page/libs/features/mas/docs/mas.js.html

### CODE REVIEW

File: libs/features/mas/commerce/libs/commerce.d.ts [hunk 1]
```diff
@@ -38,20 +38,12 @@ declare global {
         ) => T &
             PlaceholderConstants & { new (): InstanceType<T> & Placeholder };
 
-        type getLocaleSettings = (
```
Comment (3ch023) [inline libs/features/mas/commerce/libs/commerce.d.ts:41]: I'd just remove this whole file if @yesil has no obj

**Review by 3ch023 — APPROVED** [review]
lgtm

File: libs/features/mas/commerce/libs/commerce.d.ts (hunk 1 above)
Comment (npeltier) [reply libs/features/mas/commerce/libs/commerce.d.ts:41]: yeah, let's discuss this on monday but i would agree too

File: libs/features/mas/commerce/src/localeSettings.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  return {
+      country,
+      language,
+      locale: `${language}_${country}`,
+  };
+}
+
+function getLocaleSettings({ locale, country, language }) {
+  if (locale?.prefix) return getMiloLocaleSettings(locale);
+  if (typeof(locale) === 'object') locale = null;
```
Comment (3ch023) [inline libs/features/mas/commerce/src/localeSettings.js:105]: imo reassigning the parameter makes it bit confusing, i'd have a separate variable like `masLocale` that would be build out of `locale` (milo one) and returned

File: libs/features/mas/commerce/src/localeSettings.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  language = language.toLowerCase();
+
+  return {
+      country,
+      language,
+      locale: `${language}_${country}`,
+  };
+}
+
+function getLocaleSettings({ locale = {}, country = undefined, language = undefined } = {}) {
```
Comment (3ch023) [inline libs/features/mas/commerce/src/localeSettings.js:103]: we need 3 values from the consumer:
country, language, and literalsLanguage (for price literals)

Example I am Canadian french checking Spanish version of an App, so I need:
country=CA, language=FR, priceLitaralsLanguage=ES

I think it's easier consumers have only 1 way to provide commerce locale, there is no need for separate country and language

sample link for WCS where the need for locale is visible:
https://www.adobe.com/web_commerce_artifact?offer_selector_ids=MkR_-obCKAXTVvDwFzdmlC81ozQ1oxVwvC7uVbaiyB4&country=CA&locale=fr_CA&landscape=PUBLISHED&api_key=wcms-commerce-ims-ro-user-milo&language=MULT

https://www.adobe.com/web_commerce_artifact?offer_selector_ids=MkR_-obCKAXTVvDwFzdmlC81ozQ1oxVwvC7uVbaiyB4&country=CA&locale=en_CA&landscape=PUBLISHED&api_key=wcms-commerce-ims-ro-user-milo&language=MULT

see the price format difference

File: libs/features/mas/build-docs.mjs [hunk 4]
```diff
@@ -81,7 +77,17 @@ const htmlTemplate = `
...
+  <script type="module" src="../../../deps/mas/mas.js"></script>
+  <mas-commerce-service></mas-commerce-service>
+  <script type="module">
+    const params = new URLSearchParams(document.location.search);
+    const masCommerceService = document.querySelector('mas-commerce-service');
+    ['locale','language','env'].forEach((attribute) => {
+      const value = params.get(attribute);
+      if (value) masCommerceService.setAttribute(attribute, value);
+    });
+    masCommerceService.activate();
```
Comment (yesil) [inline libs/features/mas/build-docs.mjs:89]: I think ` masCommerceService.activate();` should not be needed. As soon as it is attached to DOM, it should initialize. Ideally we should create ` <mas-commerce-service></mas-commerce-service>` with the set of attributes.

File: libs/features/mas/commerce/libs/commerce.d.ts (hunk 1 above)
Comment (yesil) [reply libs/features/mas/commerce/libs/commerce.d.ts:41]: I agree, but I would like to move them to jsdoc. By just removing we will lose valuable typing information.

File: libs/features/mas/commerce/libs/commerce.d.ts [hunk 5]
```diff
@@ -608,9 +595,9 @@ export declare const getSettings: Commerce.getSettings;
...
  * init(getConfig, true);
  */
-export declare const init: Commerce.initService;
+export declare const init: Commerce.init;
 /**
  * Disposes active instance of `Commerce` service allowing it to be initialised again
  * via normal call to `init`.
  */
-export declare const reset: Commerce.resetService;
+export declare const reset: Commerce.reset;
```
Comment (yesil) [inline libs/features/mas/commerce/libs/commerce.d.ts:603]: I think we should not support resetting the service. It increasing the complexity.
with the custom element, to reset, one should remove it from DOM and add a new one.

File: libs/features/mas/commerce/src/localeSettings.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    pr: 'US_es', // not a typo, should be US
+    gt: 'GT_es',
+    cis_en: 'AZ_en',
+    cis_ru: 'AZ_ru',
+    sea: 'SG_en',
+    th_en: 'TH_en',
+    th_th: 'TH_th',
+};
+
+function getMiloLocaleSettings(locale) {
```
Comment (yesil) [inline libs/features/mas/commerce/src/localeSettings.js:87]: this should move to Milo. We should not have a Milo reference in mas-commerce.

File: libs/features/mas/commerce/src/mas-commerce-service.js [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+      });
+      resolve(this);
+    }
+
+    /**
+     * @param config, if not provided, the one from attributes of the element will be taken
+     * @param providers, if not provided, the one from registrations before activation will be taken
+     * @returns
+     */
+    async activate(config, dataProviders) {
```
Comment (yesil) [inline libs/features/mas/commerce/src/mas-commerce-service.js:136]: I strongly think that we should remove active/deactivate/reset features.
Just thing as a custom element, where it is attached and detached from DOM.

**Review by yesil — CHANGES_REQUESTED** [review]

**Comment by yesil** [comment]
@npeltier can you also rename events from `wcms:placeholder:*` to `mas:*` ?

File: libs/features/mas/commerce/src/localeSettings.js (hunk 6 above)
Comment (npeltier) [reply libs/features/mas/commerce/src/localeSettings.js:87]: i agree but was thinking of doing this later :)

File: libs/features/mas/commerce/src/mas-commerce-service.js (hunk 7 above)
Comment (npeltier) [reply libs/features/mas/commerce/src/mas-commerce-service.js:136]: ok, we can discuss that, we have a consumer willing to activate things themselves (but they can adding the component to the DOM i guess :) )

**Comment by afmicka** [comment]
@npeltier Your changes are breaking commerce in UK locale (the prices have promo code). 

With your changes, the old price is rendered only as a regular price, and optical price has a different value
- https://mwpw-159257--milo--npeltier.hlx.live/uk/drafts/nala/features/commerce/promo-placeholders
Screenshot 2024-10-15 at 12 01 54

It should be: 
- https://main--milo--adobecom.hlx.live/uk/drafts/nala/features/commerce/promo-placeholders 
Screenshot 2024-10-15 at 12 03 10

**Comment by npeltier** [comment]
> @npeltier Your changes are breaking commerce in UK locale (the prices have promo code).
> 
> With your changes, the old price is rendered only as a regular price, and optical price has a different value
> 
>     * https://mwpw-159257--milo--npeltier.hlx.live/uk/drafts/nala/features/commerce/promo-placeholders
> 
> 
> Screenshot 2024-10-15 at 12 01 54
> 
> It should be:
> 
>     * https://main--milo--adobecom.hlx.live/uk/drafts/nala/features/commerce/promo-placeholders
> 
> 
> Screenshot 2024-10-15 at 12 03 10

thanks, @afmicka, requests are the same, only difference is my PR does the request on stage for a reason i need to understand :)

**Comment by npeltier** [comment]
@yesil @3ch023 (cc @afmicka )
so basically milo's `getConfig()` on Mili's page returns 
```
env = {
  name: 'stage', 
  ims: 'stg1', 
  adobeIO: 'cc-collab-stage.adobe.io', 
  adminconsole: 'stage.adminconsole.adobe.com', 
  account: 'stage.account.adobe.com', 
  …
}
```
i guess here env.name=stage is bound to IMS, not commerce. we agree the only way to put commerce stage should be `commerce.env`, right?

**Comment by 3ch023** [comment]
> we agree the only way to put commerce stage should be `commerce.env`, right?

right.
?commerce.env=stage should result in https://www.stage.adobe.com/web_commerce_artifact_stage
?env=stage should result in https://www.adobe.com/web_commerce_artifact_stage

**Comment by npeltier** [comment]
ok @3ch023 for commerce.env, but env=stage, here it's IMS, not commerce. Typically if you look at https://main--milo--adobecom.hlx.live/uk/drafts/nala/features/commerce/promo-placeholders env=stage but it's just prod & prod uri

**Comment by 3ch023** [comment]
> ok @3ch023 for commerce.env, but env=stage, here it's IMS, not commerce. Typically if you look at https://main--milo--adobecom.hlx.live/uk/drafts/nala/features/commerce/promo-placeholders env=stage but it's just prod & prod uri

yes, I had a typo, sorry.
confirming once again ?env=stage has no effect on Commerce.
It should go to prod + prod uri.
?env=stage should result in https://www.adobe.com/web_commerce_artifact

File: libs/features/mas/commerce/src/mas-element.js [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+    version = 0;
+
+    wrapperElement;
+    
+    constructor(wrapperElement) {
+      this.wrapperElement = wrapperElement;
+    }
+
+    update() {      
+      [STATE_FAILED, STATE_PENDING, STATE_RESOLVED].forEach((state) => {
```
Comment (yesil) [inline libs/features/mas/commerce/src/mas-element.js:59]: IMO `STATE_PENDING` we should plan to remove, as it causes unnecessary updates.
the absence of resolved/rejected is by default pending state.

Comment (yesil) [reply libs/features/mas/commerce/src/mas-element.js:59]: we can do it another time.

File: libs/features/mas/commerce/src/inline-price.js [hunk 9]
```diff
@@ -1,91 +1,85 @@
...
+    TEAM_COM: ['ZA_en', 'LT_lt', 'LV_lv', 'NG_en', 'ZA_en', 'CO_es', 'KR_ko'],
+    // student
+    INDIVIDUAL_EDU: ['LT_lt', 'LV_lv', 'SA_en', 'SG_en'],
+    // school and uni
+    TEAM_EDU: ['SG_en', 'KR_ko'],
 };
 
-/** @type {Commerce.Price.PlaceholderConstructor} */
-export class HTMLPriceSpanElement extends HTMLSpanElement {
+export class InlinePrice extends HTMLSpanElement {
```
Comment (yesil) [inline libs/features/mas/commerce/src/inline-price.js:82]: I suggest to keep the Element suffix.

File: libs/features/mas/commerce/test/checkout.test.js [hunk 10]
```diff
@@ -13,15 +20,12 @@ import {
...
-
-import { mockConfig } from './mocks/config.js';
+import { useService } from '../src/utilities.js';
 import { mockFetch } from './mocks/fetch.js';
 import { mockIms, unmockIms } from './mocks/ims.js';
 import { mockLana, unmockLana } from './mocks/lana.js';
-import { mockProviders } from './mocks/providers.js';
 import { withWcs } from './mocks/wcs.js';
-import { expect, sinon } from './utilities.js';
+import { expect, sinon, initMasCommerceService, disableMasCommerceService } from './utilities.js';
```
Comment (yesil) [inline libs/features/mas/commerce/test/checkout.test.js:28]: are `initMasCommerceService ` and `disableMasCommerceService` still needed?

**Comment by npeltier** [comment]
@yesil wrt to checkout & inline price, i find it confusing to have file names not corresponding to the classes, nor the "is". That's why i just kept those two names, and reused them all over (before you had the registered class & the placeholder element with two different names which i found really confusing, now you have that class + MasElement sub item)
i'm fine renaming as you wish though, i'm passed the will to do any discussion :)

**Review by Axelcureno — APPROVED** [review]

File: libs/blocks/merch/merch.js [hunk 11]
```diff
@@ -3,7 +3,7 @@ import {
 } from '../../utils/utils.js';
 import { replaceKey } from '../../features/placeholders.js';
 
-export const CHECKOUT_LINK_CONFIG_PATH = '/commerce/checkout-link.json'; // relative to libs.
+export const CHECKOUT_LINK_CONFIG_PATH = '/commerce/checkout-link-anchor-element.json'; // relative to libs.
```
Comment (yesil) [inline libs/blocks/merch/merch.js:6]: @npeltier this is an excel file, I don't think we should rename it, especially now.

File: libs/features/mas/commerce/src/mas-commerce-service.js [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+        //root parameters
+        ['locale', 'country', 'language'].forEach((attribute) => {
+            const value = this.getAttribute(attribute);
+            if (value) {
+                config[attribute] = value;
+            }
+        });
+        //commerce parameters
+        ['checkoutWorkflowStep', 'forceTaxExclusive', 'checkoutClientId'].forEach((attribute) => {
+            const value = this.getAttribute(attribute);
```
Comment (yesil) [inline libs/features/mas/commerce/src/mas-commerce-service.js:42]: I think we should adopt html friendly attribute names: `checkout-workflow-step`, `force-tax-exclusive` `checkout-client-id`

File: libs/features/mas/mas/dist/mas.js.map
Comment (yesil) [inline libs/features/mas/mas/dist/mas.js.map:1]: might be a unintentional commit, let's remove this.

File: libs/features/mas/web-components/src/global.css.js [hunk 13]
```diff
@@ -25,6 +24,8 @@ styles.innerHTML = `
     --consonant-merch-card-cta-font-size: 15px;
 
     /* headings */
+    --merch-card-heading-xxs-font-size: 16px;
```
Comment (yesil) [inline libs/features/mas/web-components/src/global.css.js:27]: these changes that are supposed to be shipped as part of @Axelcureno's PRs, if shipped earlier, might cause regression.
I recommend to met his PRs first.

File: libs/features/mas/commerce/src/mas-commerce-service.js
[the remark this answers is not in the document]
Comment (npeltier) [reply libs/features/mas/commerce/src/mas-commerce-service.js:1]: i would rather not, a web component is an element, we know it

File: libs/features/mas/web-components/src/global.css.js (hunk 13 above)
Comment (npeltier) [reply libs/features/mas/web-components/src/global.css.js:27]: the other way around i would say

File: libs/features/mas/commerce/src/inline-price.js [hunk 14]
```diff
@@ -1,91 +1,85 @@
...
+    'ID_en',
+    'ID_in',
+    'IN_en',
+    'IN_hi',
+    'JP_ja',
+    'MY_en',
+    'MY_ms',
+    'NZ_en',
+    'TH_en',
+    'TH_th',
```
Comment (narcis-radu) [inline libs/features/mas/commerce/src/inline-price.js:57]: Locales missing

**Review by yesil — APPROVED** [review]

File: libs/blocks/merch/merch.js [hunk 15]
```diff
@@ -44,6 +45,105 @@ export const CC_SINGLE_APPS = [
...
+  ng: 'NG_en',
+  cr: 'CR_es',
+  ec: 'EC_es',
+  pr: 'US_es', // not a typo, should be US
+  gt: 'GT_es',
+  cis_en: 'AZ_en',
+  cis_ru: 'AZ_ru',
+  sea: 'SG_en',
+  th_en: 'TH_en',
+  th_th: 'TH_th',
```
[the remark this answers is not in the document]
Comment (npeltier) [reply libs/blocks/merch/merch.js:127]: this is the locales we are living with at the moment, i just moved them to a different place. Do you have an example?

**Comment by afmicka** [comment]
@npeltier 

1. `commerce.env=stage` is making requests to www.adobe.com instead of www.stage.adobe.com
https://mwpw-159257--milo--npeltier.hlx.page/drafts/nala/features/commerce/promo-placeholders?commerce.env=stage

Screenshot 2024-10-18 at 12 40 42

2. `commerce.landscape` does not work (has effect) on your branch, 
https://mwpw-159257--milo--npeltier.hlx.page/drafts/nala/features/commerce/promo-placeholders?commerce.landscape=draft

Screenshot 2024-10-18 at 12 46 36

on main branch: https://main--milo--adobecom.hlx.page/drafts/nala/features/commerce/promo-placeholders?commerce.landscape=draft
Screenshot 2024-10-18 at 12 47 43


the rest of the regression looks OK but after fixing this, it needs another round. @Roycethan could you follow-up after the fix? I will be away starting this afternoon until Thursday morning.

**Comment by Roycethan** [comment]
Regressions covered for this branch - Once above mentioned issues get fixed let me know  @npeltier @3ch023  @yesil

**Comment by Roycethan** [comment]
Reported issues fixed and - regressions covered

**Pull request merged**